tpm2tools/keys.go

```diff
@@ -168,35 +169,45 @@ func (k *Key) Close() {
 	tpm2.FlushContext(k.rw, k.handle)
 }
 
-// Seal seals the sensitive byte buffer to the provided PCRs under the owner
-// hierarchy using the SHA256 versions of the provided PCRs.
+// Seal will seal the sensitive data with the given SealingOpt.
+// SealingOpt contains some PCRs values which will bind to the secret. Those PCRs
+// will also be used to generate a ticket to certify the seal in the future.
+// SealingOpt can be nil, in which case the the auth will be empty.
 // The Key k is used as the parent key.
-func (k *Key) Seal(sensitive []byte, sel tpm2.PCRSelection) (*proto.SealedBytes, error) {
+func (k *Key) Seal(sensitive []byte, sOpt SealingOpt) (*proto.SealedBytes, error) {
 	var auth []byte
-	if len(sel.PCRs) > 0 {
-		pcrs, err := ReadPCRs(k.rw, sel)
+	var pcrList []int32
+	var pcrHash proto.HashAlgo
+	var certifyPCRs tpm2.PCRSelection
+	var err error
+
+	if sOpt != nil {
+		pcrs, err := sOpt.PCRsForSealing(k.rw)
 		if err != nil {
-			return nil, fmt.Errorf("could not read pcrs for sealing: %v", err)
+			return nil, err
 		}
 		auth = ComputePCRSessionAuth(pcrs)
-	}
 
-	sb, err := sealHelper(k.rw, k.Handle(), auth, sensitive)
+		for pcrNum := range pcrs.GetPcrs() {
+			pcrList = append(pcrList, int32(pcrNum))
+		}
+		certifyPCRs = sOpt.GetPCRSelection()
+		pcrHash = pcrs.GetHash()
+	}
+	sb, err := sealHelper(k.rw, k.Handle(), auth, sensitive, certifyPCRs)
 	if err != nil {
 		return nil, err
 	}
-	for _, pcr := range sel.PCRs {
-		sb.Pcrs = append(sb.Pcrs, int32(pcr))
-	}
-	sb.Hash = proto.HashAlgo(sel.Hash)
+	sb.Pcrs = pcrList
+	sb.Hash = pcrHash
 	sb.Srk = proto.ObjectType(k.pubArea.Type)
 	return sb, nil
 }
 
-func sealHelper(rw io.ReadWriter, parentHandle tpmutil.Handle, auth []byte, sensitive []byte) (*proto.SealedBytes, error) {
+func sealHelper(rw io.ReadWriter, parentHandle tpmutil.Handle, auth []byte, sensitive []byte, certifyPCRsSel tpm2.PCRSelection) (*proto.SealedBytes, error) {
```

Comment:
As we only call this once, should we move it into Seal?

Reply:
We can, but I think it will make the Seal function too long.